File: cpp/leetcode/lc_0121_best_time_to_buy_and_sell_stock.cpp
#include <algorithm>
#include <climits>
#include <vector>
using std::vector;

int maxProfit_2Pointers(vector<int>& prices) {
    int left = 0;
    int right = 1;
    int max = 0;

    while (right < prices.size()) {

        if (prices[left] < prices[right]) {

            int current = prices[right] - prices[left];
            max = std::max(max, current);
        } else {
            left = right;
        }

        right++;
    }

    return max;
}

int maxProfit(vector<int>& prices) {
    int min = INT_MAX;
    int max = 0;

    for (int i = 0; i < prices.size(); i++) {
        if (prices[i] < min) {
            min = prices[i];
        } else if (prices[i] - min > max) {
            max = prices[i] - min;
        }
    }
    return max;
}

File: rust/leetcode/src/lc_0121_best_time_to_buy_and_sell_stock.rs
pub fn max_profit(prices: Vec<i32>) -> i32 {
    let mut left = 0;
    let mut right = 1;
    let mut max = 0;

    while right < prices.len() {
        if prices[left] < prices[right] {
            let profit = prices[right] - prices[left];
            max = std::cmp::max(max, profit);
        } else {
            left = right;
        }
        right += 1;
    }
    max
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn max_profit_1() {
        assert_eq!(max_profit(vec![7, 1, 5, 3, 6, 4]), 5);
    }

    #[test]
    fn max_profit_2() {
        assert_eq!(max_profit(vec![7, 6, 4, 3, 1]), 0);
    }

    #[test]
    fn max_profit_3() {
        assert_eq!(max_profit(vec![2, 4, 1]), 2);
    }
}
